baseapp/baseapp.go

```diff
@@ -526,53 +510,74 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
 		ctx = ctx.WithMultiStore(msCache)
 	}
 
-	finalResult := sdk.Result{}
-	var logs []string
-	for i, msg := range msgs {
-		// Match route.
-		msgType := msg.Type()
-		handler := app.router.Route(msgType)
-		if handler == nil {
-			return sdk.ErrUnknownRequest("Unrecognized Msg type: " + msgType).Result()
-		}
+	// No need to run Msg checks when running CheckTx(), since CheckTx() only needs to validate that fees can be paid
+	// Then when DeliverTx() is called, this will run, preventing us from doing the same checks twice
+	if mode != runTxModeCheck {
 
-		result = handler(ctx, msg)
-
-		// Set gas utilized
-		finalResult.GasUsed += ctx.GasMeter().GasConsumed()
-		finalResult.GasWanted += result.GasWanted
+		// Get the Msg.
+		var msgs = tx.GetMsgs()
+		if msgs == nil || len(msgs) == 0 {
+			return sdk.ErrInternal("Tx.GetMsgs() must return at least one message in list").Result()
+		}
 
-		// Append Data and Tags
-		finalResult.Data = append(finalResult.Data, result.Data...)
-		finalResult.Tags = append(finalResult.Tags, result.Tags...)
+		for _, msg := range msgs {
+			// Validate the Msg
+			err := msg.ValidateBasic()
```

Comment:
The intent of ValidateBasic is a quick validity check to filter out garbage or mistakes in messages. It should be cheap enough that we don't really need to worry about the cost, and we should run it during CheckTx too, even before we check the fees. So this should probably go before we call the AnteHandler. That said, we may want to limit the number of messages per transaction for now